internal/command/init.go

```diff
@@ -808,6 +815,571 @@ func (c *InitCommand) getProviders(ctx context.Context, config *configs.Config,
 	return true, false, diags
 }
 
+// getProvidersFromConfig determines what providers are required by the given configuration data.
+// The method downloads any missing providers that aren't already downloaded and then returns
+// dependency lock data based on the configuration.
+// The dependency lock file itself isn't updated here.
+func (c *InitCommand) getProvidersFromConfig(ctx context.Context, config *configs.Config, upgrade bool, pluginDirs []string, flagLockfile string, view views.Init) (output bool, resultingLocks *depsfile.Locks, diags tfdiags.Diagnostics) {
+	ctx, span := tracer.Start(ctx, "install providers")
+	defer span.End()
+
+	// Dev overrides cause the result of "terraform init" to be irrelevant for
+	// any overridden providers, so we'll warn about it to avoid later
+	// confusion when Terraform ends up using a different provider than the
+	// lock file called for.
+	diags = diags.Append(c.providerDevOverrideInitWarnings())
+
+	// Collect the provider dependencies from the configuration.
+	reqs, hclDiags := config.ProviderRequirements()
+	diags = diags.Append(hclDiags)
+	if hclDiags.HasErrors() {
+		return false, nil, diags
+	}
+
+	for providerAddr := range reqs {
+		if providerAddr.IsLegacy() {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"Invalid legacy provider address",
+				fmt.Sprintf(
+					"This configuration or its associated state refers to the unqualified provider %q.\n\nYou must complete the Terraform 0.13 upgrade process before upgrading to later versions.",
+					providerAddr.Type,
+				),
+			))
```

Comment:
Just curious (i.e. I'm not implying this is wrong in any way) - what is the rationale behind not returning early here? Are we collecting this along with some other diagnostics?

Reply:
This section of code is just copied as-is. It looks like when this legacy check was first added to getProviders it had a n error check immediately underneath it, and later some extra code was inserted under it and separated it from its error check. That made it easy to lose the error check when refactoring.

I'll add a return here